README.md
# Redir [![Circle CI](https://circleci.com/gh/mesosphere/redir.svg?style=svg)](https://circleci.com/gh/mesosphere/redir)

Redir is an HTTP redirector of DNS SRV records with configurable
load-balancing strategies.

## Installing 

Pre-compiled binaries are [available](https://github.com/mesosphere/redir/releases)
for most OS and architectures. 

Alternatively, if you have Go installed:
```shell
$ go get github.com/mesosphere/redir
```

## Running
```shell
$ redir -h

usage: redir [flags]
  -addr=:8080: HTTP address to listen on
  -code=307: HTTP code to respond with
  -origin="*": HTTP CORS Origin to accept
  -param="request-path": DNS SRV name source [request-path, host-header]
  -path="/go/": HTTP path to handle
  -resolver=:53: DNS resolver addr to use
  -strategy="random": SRV RR load balancing strategy [random, round-robin]
  -timeout=1s: DNS query timeout

description:
  This program starts a CORS enabled HTTP server on a given -addr whose
  requests to -path are converted into -timeout bound DNS SRV queries against
  the -resolver. The -param defines where the name to be resolved comes from.
  Answers returned are load balanced with the given -strategy and used to respond
  to the original request with the defined -code and derived location.
```


CHANGELOG
2015-08-11: v0.0.1
  - Initial relase


handlers_test.go
package main

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/miekg/dns"
)

func TestRedirectHandler(t *testing.T) {
	c := clientFunc(func(m *dns.Msg, _ string) (*dns.Msg, time.Duration, error) {
		var r dns.Msg
		r.SetReply(m)
		r.Answer = map[string][]dns.RR{
			"_abc._tcp.domain.": []dns.RR{
				&dns.SRV{Target: "abc.domain.", Port: 5000},
				&dns.SRV{Target: "abc.domain.", Port: 5001},
				&dns.SRV{Target: "abc.domain.", Port: 5002},
				&dns.SRV{Target: "abc.domain.", Port: 5003},
			},
		}[m.Question[0].Name]
		return &r, 0, nil
	})

	for i, tt := range []struct {
		strategy
		param
		name     string
		code     int
		location string
	}{
		{roundRobin(0), path("/"), "_abc._tcp.domain", 303, "http://abc.domain.:5001"},
		{roundRobin(1), header("Host"), "_abc._tcp.domain", 303, "http://abc.domain.:5002"},
		{roundRobin(2), path("/"), "_abc._tcp.domain.", 303, "http://abc.domain.:5003"},
		{roundRobin(3), header("Host"), "_abc._tcp.domain.", 303, "http://abc.domain.:5000"},

		{random(0), path("/"), "_abc._tcp.domain", 303, "http://abc.domain.:5002"},
		{random(1), header("Host"), "_abc._tcp.domain", 303, "http://abc.domain.:5001"},
		{random(2), path("/"), "_abc._tcp.domain.", 303, "http://abc.domain.:5002"},
		{random(3), header("Host"), "_abc._tcp.domain.", 303, "http://abc.domain.:5000"},

		{roundRobin(0), path("/"), "_missing._tcp.domain.", 400, ""},
		{random(0), header("Host"), "_missing._tcp.domain.", 400, ""},
	} {
		uri := fmt.Sprintf("http://%s/%s", tt.name, tt.name)
		req, err := http.NewRequest("GET", uri, nil)
		if err != nil {
			t.Fatal(err)
		}
		rw := httptest.NewRecorder()
		h := redirectHandler(c, "/", tt.code, tt.strategy, tt.param)

		h.ServeHTTP(rw, req)
		if got, want := rw.Code, tt.code; got != want {
			t.Errorf("test #%d: got code: %d, want: %d", i, got, want)
		} else if got, want := rw.HeaderMap.Get("Location"), tt.location; got != want {
			t.Errorf("test #%d: got location: %v, want: %v", i, got, want)
		}
	}
}

// clientFunc is a function type that implements the client interface
type clientFunc func(*dns.Msg, string) (*dns.Msg, time.Duration, error)

func (f clientFunc) Exchange(m *dns.Msg, addr string) (*dns.Msg, time.Duration, error) {
	return f(m, addr)
}


handlers.go
package main

import (
	"fmt"
	"net/http"
	"strings"

	"github.com/miekg/dns"
)

// redirectHandler returns an http.Handler which resolves all requests
// of the form /path/:name and responds with an HTTP with the given code,
// pointing to the record chosen by the given strategy.
func redirectHandler(c client, resolver string, code int, st strategy, name param) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var msg dns.Msg
		msg.SetQuestion(dns.Fqdn(name(r)), dns.TypeSRV)

		if res, _, err := c.Exchange(&msg, resolver); err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
		} else if srv := st(res.Answer); srv == nil {
			http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
		} else {
			location := fmt.Sprintf("http://%s:%d", srv.Target, srv.Port)
			http.Redirect(w, r, location, http.StatusSeeOther)
		}
	})
}

// param is a function type that extracts a string parameter from an *http.Request
type param func(*http.Request) string

// path returns a param function that strips the given prefix from an http.Request
// URL Path and returns the rest.
func path(strip string) param {
	return func(r *http.Request) string {
		return r.URL.Path[len(strip):]
	}
}

// header returns a param function that returns the given header from an http.Request
func header(name string) param {
	return func(r *http.Request) string {
		if strings.ToLower(name) == "host" {
			return r.Host
		}
		return r.Header.Get(name)
	}
}


decorators.go
package main

import (
	"encoding/json"
	"io"
	"net/http"
	"strings"
	"time"

	c0rs "github.com/streadway/handy/cors"
)

// A decorator decorates an http.Handler with a layer of behaviour
type decorator func(http.Handler) http.Handler

// decorate decorates an http.Handler with all the given decorators, in order.
func decorate(h http.Handler, ds ...decorator) http.Handler {
	decorated := h
	for _, decorate := range ds {
		decorated = decorate(decorated)
	}
	return decorated
}

// logging returns a decorator which wraps an http.Handler with structured
// logging.
func logging(w io.Writer) decorator {
	logger := json.NewEncoder(w)
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			rw := capture{w, 200, nil}
			began := time.Now().UTC()
			next.ServeHTTP(&rw, r)
			_ = logger.Encode(map[string]interface{}{
				"time":           began,
				"latency_ns":     time.Since(began),
				"method":         r.Method,
				"code":           rw.code,
				"body":           string(rw.body),
				"url":            r.URL.String(),
				"host":           r.Host,
				"remote-addr":    r.RemoteAddr,
				"proto":          r.Proto,
				"content-length": r.ContentLength,
				"headers":        r.Header,
			})
		})
	}
}

// cors returns a decorator which extends an http.Handler with CORS handling.
func cors(origin string) decorator {
	return c0rs.Middleware(c0rs.Config{
		AllowOrigin: func(*http.Request) string { return origin },
	})
}

// methods returns a decorator which wraps an http.Handler with request method
// verification, responding with MethodNotAllowed if failed.
func methods(ms ...string) decorator {
	const code = http.StatusMethodNotAllowed
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			for _, method := range ms {
				if strings.ToUpper(r.Method) == strings.ToUpper(method) {
					next.ServeHTTP(w, r)
					return
				}
			}
			http.Error(w, http.StatusText(code), code)
		})
	}
}

// capture records the response code and body returned through the embedded
// ResponseWriter calls.
type capture struct {
	http.ResponseWriter
	code int
	body []byte
}

// WriteHeader captures the returned code and delegates
func (w *capture) WriteHeader(code int) {
	w.code = code
	w.ResponseWriter.WriteHeader(code)
}

// Write captures the response body and delegates
func (w *capture) Write(b []byte) (int, error) {
	w.body = b
	return w.ResponseWriter.Write(b)
}


strategies.go
package main

import (
	"math/rand"
	"sync/atomic"

	"github.com/miekg/dns"
)

// strategy represents a load balancing strategy
type strategy func([]dns.RR) *dns.SRV

// random returns a random load balancing strategy seeded with the given seed.
func random(seed int64) strategy {
	rnd := rand.New(rand.NewSource(seed))
	return func(rrs []dns.RR) *dns.SRV {
		if srvs := collect(rrs); len(srvs) > 0 {
			return srvs[rnd.Intn(len(srvs))]
		}
		return nil
	}
}

// roundRobin returns a round-robin load balancing strategy starting with the
// given robin, going through the priority sorted SRV records.
func roundRobin(robin uint64) strategy {
	return func(rrs []dns.RR) *dns.SRV {
		if srvs := collect(rrs); len(srvs) > 0 {
			return srvs[atomic.AddUint64(&robin, 1)%uint64(len(srvs))]
		}
		return nil
	}
}

// collect collects SRV records out of a dns.RR slice.
func collect(rrs []dns.RR) []*dns.SRV {
	srvs := make([]*dns.SRV, 0, len(rrs))
	for _, r := range rrs {
		if srv, ok := r.(*dns.SRV); ok {
			srvs = append(srvs, srv)
		}
	}
	return srvs
}


main.go
package main

import (
	"flag"
	"fmt"
	"net/http"
	"os"
	"runtime"
	"time"
)

const description = `
  This program starts a CORS enabled HTTP server on a given -addr whose
  requests to -path are converted into -timeout bound DNS SRV queries against
  the -resolver. The -param defines where the name to be resolved comes from. 
  Answers returned are load balanced with the given -strategy and used to respond
  to the original request with the defined -code and derived location.
`

func main() {
	runtime.GOMAXPROCS(runtime.NumCPU())

	cfg := config{addr: ":8080", resolver: ":53"}
	fs := flag.NewFlagSet("redir", flag.ContinueOnError)
	fs.StringVar(&cfg.path, "path", "/go/", "HTTP path to handle")
	fs.StringVar(&cfg.param, "param", "request-path", "DNS SRV name source [request-path, host-header]")
	fs.StringVar(&cfg.strategy, "strategy", "random", "SRV RR load balancing strategy [random, round-robin]")
	fs.StringVar(&cfg.origin, "origin", "*", "HTTP CORS Origin to accept")
	fs.IntVar(&cfg.code, "code", http.StatusTemporaryRedirect, "HTTP code to respond with")
	fs.DurationVar(&cfg.timeout, "timeout", time.Second, "DNS query timeout")
	fs.Var(&cfg.addr, "addr", "HTTP address to listen on")
	fs.Var(&cfg.resolver, "resolver", "DNS resolver addr to use")

	fs.SetOutput(os.Stderr)
	fs.Usage = func() {
		fmt.Fprintf(os.Stderr, "\nusage: redir [flags]\n")
		fs.PrintDefaults()
		fmt.Fprintf(os.Stderr, "\ndescription:%s\n", description)
	}

	if err := fs.Parse(os.Args[1:]); err != nil {
		os.Exit(1)
	}

	strategy, err := cfg.Strategy()
	if err != nil {
		fatal(fs, err)
	}

	param, err := cfg.Param()
	if err != nil {
		fatal(fs, err)
	}

	cli := newClient(cfg.timeout)
	http.Handle(cfg.path, decorate(
		redirectHandler(cli, cfg.resolver.String(), cfg.code, strategy, param),
		methods("GET"),
		cors(cfg.origin),
		logging(os.Stdout),
	))

	fmt.Fprintf(os.Stderr, "Listening on %q and redirecting %s requests\n", cfg.addr, cfg.path)
	if err := http.ListenAndServe(string(cfg.addr), nil); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}

// config holds the redir program configuration and exposes methods which
// validate it.
type config struct {
	path     string
	param    string
	strategy string
	origin   string
	code     int
	timeout  time.Duration
	addr     address
	resolver address
}

// Strategy returns the configured load balancing strategy function.
func (cfg config) Strategy() (strategy, error) {
	switch cfg.strategy {
	case "round-robin", "rr":
		return roundRobin(0), nil
	case "random", "rand":
		return random(time.Now().UnixNano()), nil
	default:
		return nil, fmt.Errorf("unsupported strategy %q", cfg.strategy)
	}
}

// Param returns the configured DNS SRV name param function.
func (cfg config) Param() (param, error) {
	switch cfg.param {
	case "request-path", "path":
		return path(cfg.path), nil
	case "host-header", "host":
		return header("Host"), nil
	default:
		return nil, fmt.Errorf("unsupported param %q", cfg.param)
	}
}

func fatal(fs *flag.FlagSet, err error) {
	fmt.Fprintln(os.Stderr, err)
	fs.Usage()
	os.Exit(1)
}
